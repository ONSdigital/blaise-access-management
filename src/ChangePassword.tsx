import React, {useState} from "react";
import {Link, Redirect, useParams} from "react-router-dom";
import {ONSPasswordInput} from "./Components/ONSPasswordInput";
import {isDevEnv} from "./Functions";
import {ONSButton} from "./Components/ONSButton";


interface Parmas {
    user: string
}


function ChangePassword() {
    // We can use the `useParams` hook here to access
    // the dynamic pieces of the URL.
    const {user}: Parmas = useParams();
    const [buttonLoading, setButtonLoading] = useState<boolean>(false);
    const [password, setPassword] = useState<string>("");
    const [confirmPassword, setConfirmPassword] = useState<string>("");
    const [message, setMessage] = useState<string>("");
    const [redirect, setRedirect] = useState<boolean>(false);

    function changePassword() {
        if (password === "") {
            setMessage("Passwords cannot be blank");
            return;
        }
        if (password !== confirmPassword) {
            setMessage("Passwords do not match");
            return;
        }

        setButtonLoading(true);
        fetch("/api/change_password/" + user, {
            "headers": {
                "password": password,
            }
        })
            .then((r: Response) => {
                if (r.status === 200) {
                    r.json()
                        .then((json) => {
                                console.log("Retrieved users list, " + json.length + " items/s");
                                isDevEnv() && console.log(json);
                                setButtonLoading(false);
                                setMessage(json);
                                setRedirect(true);
                                return;
                            }
                        ).catch(() => {
                        console.error("Unable to read json from response");
                        setMessage("Set password failed");
                        setButtonLoading(false);
                    });
                } else {
                    console.error("Failed to retrieve instrument list, status " + r.status);
                    setMessage("Set password failed");
                    setButtonLoading(false);
                }
            }).catch(() => {
                console.error("Failed to retrieve instrument list");
                setMessage("Set password failed");
                setButtonLoading(false);
            }
        );
    }

    return (
        <>
            {
                redirect && <Redirect to={{
                    pathname: "/",
                    state: {updatedPanel: {visible: true, message: "Password for user " + user + " changed", status: "success"} }
                }}/>
            }
            <p>
                <Link to={"/"}>Previous</Link>
            </p>
            <h1>Change password for user <em>{user}</em></h1>
            <p>
                {message}
            </p>
            <form onSubmit={() => changePassword()}>
                <ONSPasswordInput label={"New password"}
                                  autoFocus={true}
                                  value={password}
                                  onChange={(e) => setPassword(e.target.value)}/>
                <ONSPasswordInput label={"Confirm password"}
                                  value={confirmPassword}
                                  onChange={(e) => setConfirmPassword(e.target.value)}/>
                <ONSButton
                    label={"Save"}
                    primary={true}
                    loading={buttonLoading}
                    onClick={() => changePassword()}/>
            </form>


        </>
    );
}

export default ChangePassword;
