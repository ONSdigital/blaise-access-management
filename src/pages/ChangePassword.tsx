import React, {useState} from "react";
import {Link, Redirect, useParams} from "react-router-dom";
import {isDevEnv} from "../Functions";
import {ONSButton, ONSPanel, ONSPasswordInput} from "blaise-design-system-react-components";



interface Parmas {
    user: string
}


function ChangePassword() {
    // We can use the `useParams` hook here to access
    // the dynamic pieces of the URL.
    const {user}: Parmas = useParams();
    const [buttonLoading, setButtonLoading] = useState<boolean>(false);
    const [password, setPassword] = useState<string>("");
    const [confirmPassword, setConfirmPassword] = useState<string>("");
    const [message, setMessage] = useState<string>("");
    const [redirect, setRedirect] = useState<boolean>(false);

    function changePassword() {
        if (password === "") {
            setMessage("Passwords cannot be blank");
            return;
        }
        if (password !== confirmPassword) {
            setMessage("Passwords do not match");
            return;
        }

        setButtonLoading(true);
        fetch("/api/change_password/" + user, {
            "headers": {
                "password": password,
            }
        })
            .then((r: Response) => {
                if (r.status === 204) {
                    setButtonLoading(false);
                    setRedirect(true);
                } else {
                    console.error("Failed to retrieve instrument list, status " + r.status);
                    setMessage("Set password failed");
                    setButtonLoading(false);
                }
            }).catch(() => {
                console.error("Failed to retrieve instrument list");
                setMessage("Set password failed");
                setButtonLoading(false);
            }
        );
    }

    return (
        <>
            {
                redirect && <Redirect to={{
                    pathname: "/",
                    state: {updatedPanel: {visible: true, message: "Password for user " + user + " changed", status: "success"} }
                }}/>
            }
            <p>
                <Link to={"/"}>Previous</Link>
            </p>
            <h1>Change password for user <em>{user}</em></h1>
            <ONSPanel hidden={(message === "")} status="error">
                {message}
            </ONSPanel>
            <form onSubmit={() => changePassword()}>
                <ONSPasswordInput label={"New password"}
                                  autoFocus={true}
                                  value={password}
                                  onChange={(e) => setPassword(e.target.value)}/>
                <ONSPasswordInput label={"Confirm password"}
                                  value={confirmPassword}
                                  onChange={(e) => setConfirmPassword(e.target.value)}/>
                <ONSButton
                    label={"Save"}
                    primary={true}
                    loading={buttonLoading}
                    onClick={() => changePassword()}/>
            </form>


        </>
    );
}

export default ChangePassword;
