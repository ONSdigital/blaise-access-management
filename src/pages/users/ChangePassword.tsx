import React, { ReactElement, useState } from "react";
import { Navigate, useParams } from "react-router-dom";
import { ONSButton, ONSPanel, ONSPasswordInput } from "blaise-design-system-react-components";
import { BreadcrumbItem } from "../../../Interfaces";
import { AuthManager } from "blaise-login-react/blaise-login-react-client";
import { UserRouteParams } from "../../../Interfaces/usersPage";
import Breadcrumbs from "../../Components/Breadcrumbs";

function ChangePassword(): ReactElement {
    const { user }: UserRouteParams = useParams() as unknown as UserRouteParams;
    const [buttonLoading, setButtonLoading] = useState<boolean>(false);
    const [password, setPassword] = useState<string>("");
    const [confirmPassword, setConfirmPassword] = useState<string>("");
    const [message, setMessage] = useState<string>("");
    const [redirect, setRedirect] = useState<boolean>(false);

    function changePassword() {
        if (password === "") {
            setMessage("Passwords cannot be blank");
            return;
        }
        if (password !== confirmPassword) {
            setMessage("Passwords do not match");
            return;
        }

        setButtonLoading(true);
        const authManager = new AuthManager();
        fetch("/api/change_password/" + user, {
            "headers": Object.assign({}, {
                "password": password
            }, authManager.authHeader())
        })
        .then((r: Response) => {
            if (r.status === 204) {
                setButtonLoading(false);
                setRedirect(true);
            } else {
                setMessage("Set password failed");
                setButtonLoading(false);
            }
        }).catch(() => {
            setMessage("Set password failed");
            setButtonLoading(false);
        });
    }

    const breadcrumbList: BreadcrumbItem[] = [
        { link: "/", title: "Home" },
        { link: "/users", title: "Manage users" }
    ];

    return (
        <>
            {
                redirect && <Navigate
                    to={{ pathname: "/users" }}
                    state={{
                        updatedPanel: {
                            visible: true,
                            message: "Password for user " + user + " changed",
                            status: "success"
                        }
                    }}
                />
            }
            <Breadcrumbs BreadcrumbList={breadcrumbList} />

            <main id="main-content" className="ons-page__main ons-u-mt-no">
                <h1 className="ons-u-mb-l">Change password for user <em>{user}</em></h1>
                <ONSPanel hidden={(message === "")} status="error">
                    {message}
                </ONSPanel>
                <form onSubmit={() => changePassword()}>
                    <ONSPasswordInput label={"New password"}
                        autoFocus={true}
                        value={password}
                        onChange={(e) => setPassword(e.target.value)} />
                    <ONSPasswordInput label={"Confirm password"}
                        value={confirmPassword}
                        onChange={(e) => setConfirmPassword(e.target.value)} />
                    <ONSButton
                        label={"Save"}
                        primary={true}
                        loading={buttonLoading}
                        onClick={() => changePassword()} />
                </form>
            </main>
        </>
    );
}

export default ChangePassword;
