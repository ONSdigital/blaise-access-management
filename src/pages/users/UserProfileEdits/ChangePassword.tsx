import React, { ReactElement, useState } from "react";
import { Navigate, useLocation, useParams } from "react-router-dom";
import { ONSButton, ONSPanel, ONSPasswordInput } from "blaise-design-system-react-components";
import { BreadcrumbItem } from "../../../Interfaces";
import { AuthManager } from "blaise-login-react/blaise-login-react-client";
import { UserRouteParams } from "../../../Interfaces/usersPage";
import Breadcrumbs from "../../../Components/Breadcrumbs";
import UserSignInErrorPanel from "../../../Components/UserSignInErrorPanel";

export default function ChangePassword(): ReactElement {
    const { user: viewedUsername }: UserRouteParams = useParams() as unknown as UserRouteParams;
    const { state } = useLocation();
    const { currentUser } = state || { currentUser: null, viewedUserDetails: null };
    const [buttonLoading, setButtonLoading] = useState<boolean>(false);
    const [password, setPassword] = useState<string>("");
    const [confirmPassword, setConfirmPassword] = useState<string>("");
    const [message, setMessage] = useState<string>("");
    const [redirect, setRedirect] = useState<boolean>(false);

    const changePassword = () => {
        const sanitisedPassword = password.trim();
        const sanitisedConfirmPassword = confirmPassword.trim();

        if (sanitisedPassword === "" || sanitisedConfirmPassword === "") {
            setMessage("Passwords cannot be blank");
            return;
        }
        if (sanitisedPassword !== sanitisedConfirmPassword) {
            setMessage("Passwords do not match");
            return;
        }

        setButtonLoading(true);
        const authManager = new AuthManager();

        fetch("/api/change-password/" + viewedUsername,
            {
                "headers": Object.assign({}, {
                    "password": sanitisedPassword
                }, authManager.authHeader())
            })
            .then((r: Response) => {
                if (r.status === 204) {
                    setButtonLoading(false);
                    setRedirect(true);
                } else {
                    setMessage("Set password failed");
                    setButtonLoading(false);
                }
            }).catch(() => {
                setMessage("Set password failed");
                setButtonLoading(false);
            });
    };

    const breadcrumbList: BreadcrumbItem[] = [
        { link: "/", title: "Home" },
        { link: "/users/", title: "Manage users" },
        { link: `/users/${viewedUsername}`, title: "View user", state: { currentUser } }
    ];

    if (!currentUser) {
        return (<UserSignInErrorPanel/>);
    }

    return (
        <>
            {
                redirect && <Navigate
                    to={{ pathname: `/users/${viewedUsername}` }}
                    state={{
                        currentUser,
                        updatedPanel: {
                            visible: true,
                            message: "Password changed for user called " + viewedUsername,
                            status: "success"
                        }
                    }}
                />
            }
            <Breadcrumbs BreadcrumbList={breadcrumbList} />

            <main id="main-content" className="ons-page__main ons-u-mt-no">
                <h1 className="ons-u-mb-l">Change password for user <em>{viewedUsername}</em></h1>
                <ONSPanel hidden={(message === "")} status="error">
                    {message}
                </ONSPanel>
                <form onSubmit={() => changePassword()}>
                    <ONSPasswordInput label={"New password"}
                        inputId={"new-password"}
                        autoFocus={true}
                        value={password}
                        onChange={(e) => setPassword(e.target.value)} />
                    <ONSPasswordInput label={"Confirm password"}
                        inputId={"confirm-password"}
                        value={confirmPassword}
                        onChange={(e) => setConfirmPassword(e.target.value)} />
                    <ONSButton
                        label={"Save"}
                        primary={true}
                        loading={buttonLoading}
                        onClick={() => changePassword()} />
                </form>
            </main>
        </>
    );
}