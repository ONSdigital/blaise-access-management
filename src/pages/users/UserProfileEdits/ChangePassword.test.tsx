import React from "react";
import { render, cleanup } from "@testing-library/react";
import "@testing-library/jest-dom";
import { MemoryRouter, useParams } from "react-router-dom";
import userEvent from "@testing-library/user-event";
import { act } from "react-dom/test-utils";
import ChangePassword from "./ChangePassword";

jest.mock("react-router-dom", () => ({
    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
    // @ts-expect-error
    ...jest.requireActual("react-router-dom"),
    useParams: jest.fn()
}));

jest.mock("blaise-login-react/blaise-login-react-client", () => ({
    AuthManager: jest.fn().mockImplementation(() => ({
        authHeader: () => ({
            Authorization: process.env.MOCK_AUTH_TOKEN
        })
    }))
}));

global.fetch = jest.fn(() =>
    Promise.resolve({
        status: 204,
        json: () => Promise.resolve({ message: "Password changed successfully" })
    })
) as jest.Mock;

const mockUserDetails = {
    name: "testUser"
};

const mockState = {
    pathname: `/users/${mockUserDetails.name}/change-password`,
    state: { currentUser: "currentUser" }
};

beforeEach(() => {
    (fetch as jest.Mock).mockClear();
    (useParams as jest.Mock).mockReturnValue({ user: mockUserDetails.name });
});

afterEach(() => {
    jest.clearAllMocks();
    cleanup();
});

describe("ChangePassword Component", () => {
    it("matches the snapshot", async () => {
        const { asFragment } = render(
            <MemoryRouter initialEntries={[mockState]}>
                <ChangePassword />
            </MemoryRouter>
        );

        expect(asFragment()).toMatchSnapshot();
    });

    it("displays error message when passwords are empty", async () => {
        const { findByText, getByLabelText, getByText } = render(
            <MemoryRouter initialEntries={[mockState]}>
                <ChangePassword />
            </MemoryRouter>
        );

        const newPasswordInput = getByLabelText("New password");
        const confirmPasswordInput = getByLabelText("Confirm password");
        const saveButton = getByText("Save");

        act(() => {
            userEvent.type(newPasswordInput, "");
            userEvent.type(confirmPasswordInput, "");
            userEvent.click(saveButton);
        });

        expect(await findByText(/Passwords cannot be blank/i)).toBeVisible();
    });

    it("displays error message when passwords do not match", async () => {
        const { findByText, getByLabelText, getByText } = render(
            <MemoryRouter initialEntries={[mockState]}>
                <ChangePassword />
            </MemoryRouter>
        );

        const newPasswordInput = getByLabelText("New password");
        const confirmPasswordInput = getByLabelText("Confirm password");
        const saveButton = getByText("Save");

        act(() => {
            userEvent.type(newPasswordInput, "password123");
            userEvent.type(confirmPasswordInput, "password321333");
            userEvent.click(saveButton);
        });

        expect(await findByText(/Passwords do not match/i)).toBeVisible();
    });

    it("calls fetch with correct parameters upon form submission with matching passwords that remove any trailing whitespaces", async () => {
        const { getByLabelText, getByText } = render(
            <MemoryRouter initialEntries={[mockState]}>
                <ChangePassword />
            </MemoryRouter>
        );

        const newPasswordInput = getByLabelText("New password");
        const confirmPasswordInput = getByLabelText("Confirm password");
        const saveButton = getByText("Save");

        // Wait for state update
        act(() => {
            userEvent.type(newPasswordInput, "password123  ");
            userEvent.type(confirmPasswordInput, "password123       ");
            userEvent.click(saveButton);
        });

        expect(fetch).toHaveBeenCalledWith("/api/change-password/testUser", {
            headers: {
                Authorization: process.env.MOCK_AUTH_TOKEN,
                password: "password123"
            }
        });
    });

    it("calls fetch with correct parameters upon form submission with matching passwords", async () => {
        const { getByLabelText, getByText, findByText } = render(
            <MemoryRouter initialEntries={[mockState]}>
                <ChangePassword />
            </MemoryRouter>
        );

        const newPasswordInput = getByLabelText("New password");
        const confirmPasswordInput = getByLabelText("Confirm password");
        const saveButton = getByText("Save");

        // Wait for state update
        act(() => {
            userEvent.type(newPasswordInput, "password123");
            userEvent.type(confirmPasswordInput, "password123");
            userEvent.click(saveButton);
        });

        expect(fetch).toHaveBeenCalledWith("/api/change-password/testUser", {
            headers: {
                Authorization: process.env.MOCK_AUTH_TOKEN,
                password: "password123"
            }
        });
    });
});