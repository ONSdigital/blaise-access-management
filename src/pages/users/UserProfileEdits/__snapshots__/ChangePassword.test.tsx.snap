// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChangePassword Component matches the snapshot 1`] = `
<DocumentFragment>
  <nav
    aria-label="Breadcrumb"
    class="ons-breadcrumb"
  >
    <ol
      class="ons-breadcrumb__items ons-u-fs-s"
    >
      <li
        class="ons-breadcrumb__item"
        id="breadcrumb-0"
      >
        <a
          class="ons-breadcrumb__link"
          href="/"
        >
          Home
        </a>
        <svg
          class="ons-svg-icon"
          focusable="false"
          viewBox="0 0 8 13"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M5.74,14.28l-.57-.56a.5.5,0,0,1,0-.71h0l5-5-5-5a.5.5,0,0,1,0-.71h0l.57-.56a.5.5,0,0,1,.71,0h0l5.93,5.93a.5.5,0,0,1,0,.7L6.45,14.28a.5.5,0,0,1-.71,0Z"
            transform="translate(-5.02 -1.59)"
          />
        </svg>
      </li>
      <li
        class="ons-breadcrumb__item"
        id="breadcrumb-1"
      >
        <a
          class="ons-breadcrumb__link"
          href="/users/"
        >
          Manage users
        </a>
        <svg
          class="ons-svg-icon"
          focusable="false"
          viewBox="0 0 8 13"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M5.74,14.28l-.57-.56a.5.5,0,0,1,0-.71h0l5-5-5-5a.5.5,0,0,1,0-.71h0l.57-.56a.5.5,0,0,1,.71,0h0l5.93,5.93a.5.5,0,0,1,0,.7L6.45,14.28a.5.5,0,0,1-.71,0Z"
            transform="translate(-5.02 -1.59)"
          />
        </svg>
      </li>
      <li
        class="ons-breadcrumb__item"
        id="breadcrumb-2"
      >
        <a
          class="ons-breadcrumb__link"
          href="/users/testUser"
        >
          View user
        </a>
        <svg
          class="ons-svg-icon"
          focusable="false"
          viewBox="0 0 8 13"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M5.74,14.28l-.57-.56a.5.5,0,0,1,0-.71h0l5-5-5-5a.5.5,0,0,1,0-.71h0l.57-.56a.5.5,0,0,1,.71,0h0l5.93,5.93a.5.5,0,0,1,0,.7L6.45,14.28a.5.5,0,0,1-.71,0Z"
            transform="translate(-5.02 -1.59)"
          />
        </svg>
      </li>
    </ol>
  </nav>
  <main
    class="ons-page__main ons-u-mt-no"
    id="main-content"
  >
    <h1
      class="ons-u-mb-l"
    >
      Change password for user 
      <em>
        testUser
      </em>
    </h1>
    <div
      class="ons-panel ons-panel--error ons-panel--no-title  ons-u-mt-m"
      hidden=""
    >
      <div
        class="ons-panel__body "
      />
    </div>
    <form>
      <p
        class="ons-field"
      >
        <label
          class="ons-label"
          for="password"
        >
          New password
        </label>
        <span
          class="ons-checkbox ons-checkbox--toggle"
        >
          <input
            autocomplete="new-password"
            class="ons-checkbox__input"
            id="password-toggle"
            name="show-password"
            type="checkbox"
          />
          <label
            class="ons-checkbox__label "
            for="password-toggle"
            id="password-toggle-label"
          >
            Show password
          </label>
        </span>
        <input
          class="ons-input ons-input--text ons-input-type__input ons-u-mt-xs"
          data-testid="login-password-input"
          id="password"
          type="password"
          value=""
        />
      </p>
      <p
        class="ons-field"
      >
        <label
          class="ons-label"
          for="password"
        >
          Confirm password
        </label>
        <span
          class="ons-checkbox ons-checkbox--toggle"
        >
          <input
            autocomplete="new-password"
            class="ons-checkbox__input"
            id="password-toggle"
            name="show-password"
            type="checkbox"
          />
          <label
            class="ons-checkbox__label "
            for="password-toggle"
            id="password-toggle-label"
          >
            Show password
          </label>
        </span>
        <input
          class="ons-input ons-input--text ons-input-type__input ons-u-mt-xs"
          data-testid="login-password-input"
          id="password"
          type="password"
          value=""
        />
      </p>
      <button
        class="ons-btn "
        data-testid="button"
        type="button"
      >
        <span
          class="ons-btn__inner"
        >
          Save
        </span>
      </button>
    </form>
  </main>
</DocumentFragment>
`;
